feature/security/gnsi/credentialz/tests/host_certificates/host_certificates_test.go

```diff
@@ -0,0 +1,308 @@
+	if err != nil {
+		t.Fatalf("failed fetching fetching credentialz public keys, error: %s", err)
+	}
+
+	if len(response.PublicKeys) < 1 {
+		return nil
+	}
+
+	return response.PublicKeys[0].PublicKey
+}
+
+func signPublicKeyWithCAKey(t *testing.T, dir string) {
+	cmd := exec.Command(
+		"ssh-keygen",
+		"-s", "ca", // sign using this ca key
+		"-I", "dut", // key identity
+		"-h",                 // create host (not user) certificate
+		"-n", "dut.test.com", // principal(s)
+		"-V", "+52w", // validity
+		dutPublicKeyFilename,
+	)
+	cmd.Dir = dir
+
+	err := cmd.Run()
+	if err != nil {
+		t.Fatalf("failed signing dut public key with ca, error: %s", err)
+	}
+}
+
+func loadCertificate(t *testing.T, dut *ondatra.DUTDevice, dir string) {
+	certificateContents, err := os.ReadFile(fmt.Sprintf("%s/%s", dir, dutCertFilename))
+	if err != nil {
+		t.Fatalf("failed reading host signed certificate, error: %s", err)
+	}
+
+	rotateHostParameters(
+		t,
+		dut,
+		[]*credentialz.ServerKeysRequest_AuthenticationArtifacts{
+			{
+				Certificate: certificateContents,
+			},
+		},
+	)
+}
+
+func assertReturnedHostKeyIsCorrect(t *testing.T, dir, addr string) {
+	certificateContents, err := os.ReadFile(fmt.Sprintf("%s/%s", dir, dutCertFilename))
+	if err != nil {
+		t.Fatalf("failed reading host signed certificate, error: %s", err)
+	}
+
+	hostCertificate, _, _, _, err := ssh.ParseAuthorizedKey(certificateContents)
+	if err != nil {
+		t.Fatalf("failed parsing host certificate authorized (cert)key: %s", err)
+	}
+
+	var failed bool
+
+	_, err = ssh.Dial(
+		"tcp",
+		fmt.Sprintf("%s:22", addr),
+		&ssh.ClientConfig{
+			User: "admin",
+			Auth: []ssh.AuthMethod{},
+			HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
+				if !reflect.DeepEqual(hostCertificate, key) {
+					failed = true
+				}
+
+				return nil
+			},
+		},
+	)
+	if err == nil {
+		t.Fatal("dial ssh succeeded, but we expected failure")
+	}
+
+	if failed {
+		t.Fatalf(
+			"host presented key (cert) that does not match expected host certificate",
+		)
+	}
+}
+
+func assertTelemetry(t *testing.T, dut *ondatra.DUTDevice) {
+	sshServer := gnmi.Get(t, dut, gnmi.OC().System().SshServer().State())
+
+	actualHostCertificateVersion := sshServer.GetActiveHostCertificateVersion()
+
+	if actualHostCertificateVersion != hostCertificateVersion {
+		t.Fatalf(
+			"telemetry reports host certificate version is not correct\n\tgot: %s\n\twant: %s",
+			actualHostCertificateVersion, hostCertificateVersion,
+		)
+	}
+
+	actualHostCertificateCreatedOn := sshServer.GetActiveHostCertificateCreatedOn()
+
+	if time.Unix(int64(hostCertificateCreatedOn), 0).Compare(time.Unix(0, int64(actualHostCertificateCreatedOn))) != 0 {
+		t.Fatalf(
+			"telemetry reports host certificate created on is not correct\n\tgot: %d\n\twant: %d",
+			actualHostCertificateCreatedOn, hostCertificateCreatedOn,
+		)
+	}
+}
+
+func getDutAddr(t *testing.T, dut *ondatra.DUTDevice) string {
```

Comment:
i get the feeling (as I said above) that everything down to and including this function i've read in 2 the Credentialz PRs.

Please let's move all of this sort of thing to a utility library we cause write/update 1x instead of N/x across all of the codebase?

Reply:
same as previous comment